### DISCUSSION TITLE
Scripting with `page counter`

### DISCUSSION THREAD

**Discussion by DannySeidel** [typst/typst#4342] [General]
### Description

Hi typst team,

I'm currently trying to set up separate numbering for the bibliography/appendix section and the preceding main section of my typst template `supercharged-dhbw`.

The goal is to use alphabetical numbering in the bibliography/appendix section of the template. When I now do

```typst
set page(numbering: "1 of 1")

main section

set page(numbering: "a")
counter(page).update(1)

bibliography
appendix
```

my numbering gets correctly set to `"a"` but the previous part having the numbering style `"1 of 1"` now shows the page count of the section with the `"a"` numbering because `counter(page).final()` starts to count from `0` again.

To make it clear for you, here's an example:

Currently, the numberings for 5 pages with the style `"1 of 1"` and 2 pages following with the style `"a"` look like this:
1 of 2
2 of 2
3 of 2
4 of 2
5 of 2
a
b

What I want:
1 of 5
2 of 5
3 of 5
4 of 5
5 of 5
a
b

I'm not entirely sure if this behavior is expected, but if it is, how do I achieve the preceding section to show the number of pages it contains (in my case "... of 5")? If the behavior is expected, it would be great if it were possible to define multiple independent page counters.

### Operating system

macOS

### Typst version

- [X] I am using the latest version of Typst

**Comment by laurmaedje** [comment]
This is working as expected. `final()` is really just the value at the end of the document. If you want independent counters, you could try adding something labelled before the reset and retrieving the counter at that label instead.

**Reply by DannySeidel** [reply]
I just tried to do what you suggested. I gave the heading of the bibliography a label `<bib>` and set the page count like this:

```typst
context {
  let page-count = counter(page).at(<bib>)

  set page(
    numbering: [1 of #page-count]
  )
  counter(page).update(1)
}
```

But I'm honestly not sure how to pass the label to the `at()` function and how to then pass the value to numbering.

**Reply by laurmaedje** [reply]
Here is a small example of what I mean:
```typ
#set page("a8")
#set page(
  numbering: (n, _) => [
    #n of #context numbering("1", ..counter(page).at(<end>))
  ]
)

#lorem(100)

#metadata(none) <end>
#set page(numbering: "A")
#counter(page).update(1)
#lorem(30)
```
Hope that helps.

**Reply by DannySeidel** [reply]
First of all, thanks for the suggestion 🙏🏻

I tried your suggestion but didn't get it to work. When I do:
```typst
set page(
  numbering: (n, _) => [
    #n of #context numbering("1", ..counter(page).at(<end>))
  ],
)
```
I get:
```
error: missing argument: pattern parameter
      ┌─ @local/supercharged-dhbw:1.3.2/lib.typ:247:17
       │
247 │       numbering: (n, _) => [
```

I don't know how to pass the pattern here correctly. As I understand it, the variable `n` is the pattern, but where should the pattern come from?

I also tried this:
```typst
set page(
  numbering: [1 of #str(context counter(page).at(<end>).first())],
)
```
But this try also failed:
```
error: expected integer, float, version, bytes, label, type, or string, found content
      ┌─ @local/supercharged-dhbw:1.3.2/lib.typ:250:28
       │
250 │       numbering: [1 of #str(context counter(page).at(<end>).first())],
```

Is there any way to easier debug this because the errors don't really help me here.

**Reply by laurmaedje** [reply]
Sorry, I should have written a more resilient version. The pattern parameter is `_` and the problem is that the numbering is called with just a single page number instead of a pair. Replacing `_` with `..` would fix this particular issue.

However, upon further consideration, having a `1 of x` as `#set page(numbering: ..)` is still problematic because you'll, for instance, also see the "of x" part in the outline. I would recommend having just `#set page(numbering: "1")` and adding a header or footer manually with the "of x" part in there. Alternatively, you can change the numbering conditionally based on whether the second parameter exists (with a parameter `..sink`), but the other thing is simpler.

Something like:
```typ
#set page(
  numbering: "1",
  footer: context align(center, numbering(
    "1 of 1", 
    ..counter(page).get(),
    ..counter(page).at(<end>),
  ))
)
```

**Reply by DannySeidel** [reply]
The numbering style `1 of x` is only used in the main section. The outline and other pages before the introduction use Roman numbering, so this is no problem.

Thanks a lot for your help!

**Comment by DannySeidel** [comment]
As general feedback for your documentation, I think for creating standard documents and doing some basic formatting it is excellent. Still, I would sometimes like more examples when it comes to more complex scripting and templates. That would really help me in cases like this.

**Reply by laurmaedje** [reply]
Yeah, there's for sure more work to be done on the docs side!
